internal/pkg/gateway/api.go

```diff
@@ -71,25 +74,24 @@ func (gs *Server) Evaluate(ctx context.Context, request *gp.EvaluateRequest) (*g
 		gs.logger.Debugw("Sending to peer:", "channel", channel, "chaincode", chaincodeID, "MSPID", endorser.mspid, "endpoint", endorser.address)
 
 		pr, err := endorser.client.ProcessProposal(ctx, signedProposal)
-		if err != nil {
-			logger.Debugw("Evaluate call to endorser failed", "chaincode", chaincodeID, "channel", channel, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "error", err)
-			errDetails = append(errDetails, errorDetail(endorser.endpointConfig, err))
-			gs.registry.removeEndorser(endorser)
-			endorser = plan.retry(endorser)
+		success, message, retry, remove := gs.responseStatus(pr, err)
+		if success {
+			response = pr.Response
+		} else {
+			logger.Debugw("Evaluate call to endorser failed", "chaincode", chaincodeID, "channel", channel, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "error", message)
+			errDetails = append(errDetails, errorDetail(endorser.endpointConfig, message))
+			if remove {
+				gs.registry.removeEndorser(endorser)
+			}
+			if retry {
+				endorser = plan.retry(endorser)
+			} else {
+				return nil, newRpcError(codes.Aborted, "evaluate call to endorser returned error: "+message, errDetails...)
+			}
```

Comment:
Can't this and the next return be combined? Why does this one return message and errDetails, and the latter return only errDetails?
Regardless of whether we don't retry, or we've exhausted retries, shouldn't response be identical?